feat: add import and export rules

susnux opened the pull request:
* faster alternative for https://github.com/nextcloud-libraries/eslint-config/pull/973

- Enforce file extensions
- Enforce consistent named import / export sorting
- Enforce consistent module sorting

Do not use eslint-plugin-import(-x) because it is quite slow and has many issues with parsing different formats like vue script-setup in Typescript and similar.

ShGKme reviewed (commented):
Works well.

One small problem - it keeps existing empty lines between imports even after working.

ShGKme commented on lib/configs/imports.ts:
```diff
@@ -0,0 +1,116 @@
+					'error',
+					{
+						type: 'natural',
+						newlinesBetween: 'ignore',
```
I guess, because of this

Antreesy replied:
yes, works as expected with 'never'

susnux replied:
:heavy_check_mark:

ShGKme commented:
```ts
import type { NcSelectOption } from '../composables/useNcSelectModel.ts'

import { t } from '@nextcloud/l10n'
import NcButton from '@nextcloud/vue/components/NcButton'
import NcCheckboxRadioSwitch from '@nextcloud/vue/components/NcCheckboxRadioSwitch'
import NcNoteCard from '@nextcloud/vue/components/NcNoteCard'
import NcTextField from '@nextcloud/vue/components/NcTextField'
import { storeToRefs } from 'pinia'
import { computed, ref } from 'vue'
import { ZOOM_MAX, ZOOM_MIN } from '../../../constants.js'
import { useNcSelectModel } from '../composables/useNcSelectModel.ts'
import { useAppConfigStore } from './appConfig.store.ts'
import { useAppConfigValue } from './useAppConfigValue.ts'
import SettingsFormGroup from './components/SettingsFormGroup.vue'
import SettingsSelect from './components/SettingsSelect.vue'
import SettingsSubsection from './components/SettingsSubsection.vue'
import IconBellRingOutline from 'vue-material-design-icons/BellRingOutline.vue'
import IconCardAccountPhoneOutline from 'vue-material-design-icons/CardAccountPhoneOutline.vue'
import IconMagnify from 'vue-material-design-icons/Magnify.vue'
import IconMinus from 'vue-material-design-icons/Minus.vue'
import IconPhoneRingOutline from 'vue-material-design-icons/PhoneRingOutline.vue'
import IconPlus from 'vue-material-design-icons/Plus.vue'
import IconRestore from 'vue-material-design-icons/Restore.vue'
import IconThemeLightDark from 'vue-material-design-icons/ThemeLightDark.vue'
import IconVolumeHigh from 'vue-material-design-icons/VolumeHigh.vue'
```

IMHO:
- External modules `vue-material-design-icons` should not go after `internal` even they are `.vue`
- Would be nice to have imports like `from 'vue'` on top of `from '@nextcloud/*'`
- Would be nice to have `@nextcloud/vue` components close to all the other components
- I'd also put styles to the end to side effects

Antreesy commented on lib/configs/imports.ts:
```diff
@@ -0,0 +1,116 @@
+								'external',
+								'object',
+							],
+							// everything else which is everything internal
+							'unknown',
+							// import style from 'my.module.css'
+							'style',
+							// Vue components
+							'vue',
```
Agree with Grigorii (but also because I used to this order:
```suggestion
							// Imports from @nextcloud ecosystem
							'nextcloud',
							// Imports from @nextcloud/vue library
							'nextcloud-vue',
							// Vue components
							'vue',
							// everything else which is everything internal
							'unknown',
							// import style from 'my.module.css'
							'style',
```

Antreesy commented on lib/configs/imports.ts:
```diff
@@ -0,0 +1,116 @@
+						],
+						customGroups: {
+							value: {
+								vue: '\\.vue$',
```
```suggestion
								nextcloud: '^@nextcloud.*',
								'nextcloud-vue': '^@nextcloud\/vue.*',
								vue: '\\.vue$',
```

susnux commented:
> External modules vue-material-design-icons should not go after internal even they are .vue

Fixed :white_check_mark: 

> Would be nice to have @nextcloud/vue components close to all the other components

Fixed :white_check_mark: 

> I'd also put styles to the end to side effects

Fixed :white_check_mark: 

> Would be nice to have imports like from 'vue' on top of from '@nextcloud/*'

What is the use case? Its just a namespace, like every other, I would consider it really weird to see:
```ts
import { foo } from '@foo/bar'
import { computed } from 'vue'
import { t } from '@nextcloud/l10n'
```

instead of consistent alphabetical ordering of normal modules¹:
```ts
import { foo } from '@foo/bar'
import { t } from '@nextcloud/l10n'
import { computed } from 'vue'
```

¹ I mean like `type` imports or Vue components are a special kind of modules one is just type information that will be stripped away and the other is UI components for use in the template, compared to everything else which is logic code that will be used.

If I import that code form one namespace or another makes no difference for reading the code, it does not provide any information except that is a Nextcloud maintained library but what does that help you in your flow?

---

TL;DR: I am fine with it if others agree with that sorting but I would prefer only sort by module type rather than package maintainer.

ShGKme commented on lib/configs/codeStyle.ts:
```diff
@@ -1,17 +1,16 @@
+import type { ConfigOptions } from '../types.d.ts'
 /*!
  * SPDX-FileCopyrightText: 2025 Nextcloud GmbH and Nextcloud contributors
  * SPDX-License-Identifier: AGPL-3.0-or-later
  */
 import type { Linter } from 'eslint'
```
Is there a way to avoid putting an SPDX header between imports?

susnux replied:
Good question. It makes sense to keep comments where there were as they could be for that import.
I will have a look if we can fix this.

ShGKme replied:
What about adding a rule that searches for the SPDX comment, checks for it's existence, and moves it to the top? And, maybe, checks the format.

ShGKme replied:
Actually, the solution is quite simple.
There should be an empty line between the SPDX header and the first import. Otherwise this comment is the import comment, not a file comment.

So, I'd call it a local code problem, not the config problem.

ShGKme commented:
Import inside `type` imports seems lexicographical

```ts
import type { BuildConfig } from '../build/BuildConfig.types.ts'
import type { systemInfo } from './app/system.utils.ts'
import type { n, t } from '@nextcloud/l10n'
```

ShGKme commented:
It seems to make linting extremely slow...

In Talk Desktop (quite a small repo):
- Before: ~4 seconds
- After: ~112 seconds

ShGKme commented:
More specifically, it's `customGroups`. Even **an empty list** of custom groups makes everything slow
```
customGroups: {
 	value: {
	},
},
```

ShGKme commented:
@susnux Anyway, it is fixed in:
- https://github.com/azat-io/eslint-plugin-perfectionist/pull/528/

Just tested, works well.

ShGKme commented on lib/configs/imports.ts:
```diff
@@ -0,0 +1,122 @@
+							],
+							'localVue',
+							{ newlinesBetween: 'always' },
+							// import style from 'my.module.css'
+							'style',
+							// side effect only: import 'sideeffect.js'
+							'side-effect',
```
To have JS, then CSS, not JS - CSS - JS again

```suggestion
							// side effect only: import 'sideeffect.js'
							'side-effect',
							// import style from 'my.module.css'
							'style',
```

susnux replied:
done

susnux commented:
Fixed performance issue by using the new config format.

susnux commented:
> Import inside type imports seems lexicographical

resolved

ShGKme commented on lib/configs/imports.ts:
```diff
@@ -0,0 +1,127 @@
+							// type first
+							'external-type',
+							'type',
+							{ newlinesBetween: 'always' },
```
Maybe remove these empty lines as well?

ShGKme replied:
So imports are a single block, separated by an empty line from the remaining module code.
Type imports are separated visually via `import type` prefix anyway.

susnux replied:
```suggestion
							{ newlinesBetween: 'ignore' },
```

Maybe?

ShGKme reviewed (approved):
Works well now

Antreesy commented on lib/configs/imports.ts:
```diff
@@ -0,0 +1,127 @@
+								'external',
+								'object',
+							],
+							// everything else which is everything internal
+							'unknown',
+							// Vue components
+							'vue', // external modules (e.g. nextcloud-vue)
```
So we decided to go with this? external .vue components are now listed after internal non-vue modules
<img width="499" alt="image" src="https://github.com/user-attachments/assets/9f3aa2e1-7f92-4903-b35b-689ddbd22bb0" />

susnux replied:
@Antreesy one problem here is that you use the old imports they are not recognized as vue components. (The `/dist/Components`).
Instead use `/components/Nc..`.

susnux replied:
I would expect:

```
import type ...
import foo from 'bar'
import baz from './local'
import NcActions from '@nextcloud/vue...'
import SomeComponent from './SomeComponent.vue'

import 'other-side-effect'
```

ShGKme replied:
From the config it is:
- external
- internal
- vue external
- vue internal

Which is indeed different from the proposal above

susnux replied:
It was always like that here :eyes: 

- Typescript (external | internal)
- Javascript (external | internal)
- Vue (external | internal)
- Styles (external | internal)

ShGKme replied:
> Typescript (external | internal)\nJavascript (external | internal)

Does it actually separate TS from JS, or types from modules?

ShGKme replied:
> Otherwise why should you sort vue independently if you mix it with Javascript?

To not break it in the middle, because it can be in a single batch without mixing externals with internals

susnux replied:
> Does it actually separate TS from JS, or types from modules?

I do not mean `.ts` files I meat types.

- Types are stripped away so that block is only for type checking.
- Javascript is the code that stays so its the most important block.
- Vue files are components, you do not care in `<script>` about them so they do not need to interrupt reading the script sources. They are only relevant for the `<template>`.

Antreesy reviewed (approved):
Gonna be painful to get used to it

ShGKme commented:
> > External modules vue-material-design-icons should not go after internal even they are .vue
> Fixed ✅

This is why I thought, it was fixed, and externals don't go after internals

susnux commented:
Well then it was a misunderstanding. Currently there is really mixed sorting in e.g. server, but for me it seemed to be agreed that Vue components are sorted extra as they belong to `<template>` rather than `<script>`.
Thats where this sorting was coming from.

In most components this leads to
```js
import { computed } from 'vue'
import { getSomething } from '../../utils/myUtil.ts'
import NcButton from '@nextcloud/vue/components/NcButton'
import MyComponent from './MyComponent.vue'
```

Which I thought is more common than

```js
import { computed } from 'vue'
import NcButton from '@nextcloud/vue/components/NcButton'
import MyComponent from './MyComponent.vue'
import { getSomething } from '../../utils/myUtil.ts'
```

So it was about seeing directly what are components vs seeing what origins from where.
But I do not mind a different sorting if it is agreed upon at least its consistent.

ShGKme commented:
I'd continue discussion in https://github.com/nextcloud-libraries/eslint-config/pull/1002 to have it in one place
